qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        ]
+        V_ik = np.c_[V_ik, V_guess]
+        # Build `diagonal` of local Hamiltonian
+        D_i = self.Hamiltonian_manager.D_i()
+
+        n_newvecs = dim_S  # No. of vectors added is initial subspace dimension
+        converged = False
+        restart = 0
+        for k in range(1, max_iter):
+            if self.rank == 0:
+                print(
+                    f"Process rank: {self.rank}, Iterate: {k}, Working subspace dimension: {dim_S}"
+                )
+            # Gather full trial vectors added during previous iteration on each rank
+            V_new = np.zeros((n, n_newvecs), dtype="float")
+            V_inew = np.array(V_ik[:, -n_newvecs:], dtype="float")
+            self.comm.Allgatherv(
+                [V_inew, MPI.DOUBLE],
+                [
+                    V_new,
+                    n_newvecs * np.array(self.distribution),
+                    n_newvecs * np.array(self.offsets),
+                    MPI.DOUBLE,
+                ],
+            )
+            # Compute new columns of W_ik, W_inew = H_i * V_new
+            if driven_by == "explicit":
+                W_inew = self.Hamiltonian_manager.explicit_matrix_product(V_new)
+            elif driven_by == "implicit":
+                W_inew = self.Hamiltonian_manager.implicit_matrix_product(V_new)
+            else:
+                raise NotImplementedError
+            W_ik = np.c_[W_ik, W_inew]
+
+            # TODO: Don't need to reduce full matrix at each step. Just new columns/rows
+            # Compute rank computes partial update to the projected Hamiltonian S_k
+            if (k == 1) or (
+                restart == 1
+            ):  # If first iterate (or following a restart), need to compute full S_k explicitly
+                S_ik = np.dot(V_ik.T, W_ik)
+                restart = 0
+            else:  # Else, append new rows & columns
+                S_inew_c = np.dot(V_ik[:, :-n_newvecs].T, W_inew)
+                S_ik = np.c_[S_ik, S_inew_c]
+                S_inew_r = np.c_[np.dot(V_inew.T, W_ik[:, :-n_newvecs]), np.dot(V_inew.T, W_inew)]
+                S_ik = np.r_[S_ik, S_inew_r]
+            # Reduce contributions and form new S_k
+            S_k = None  # Initialize
+            if self.rank == 0:
+                S_k = np.zeros((dim_S, dim_S), dtype="float")
+            self.comm.Reduce([S_ik, MPI.DOUBLE], [S_k, MPI.DOUBLE])
+            if self.rank == 0:  # Master diagonalizes S_k
+                L_k, Y_k = np.linalg.eigh(S_k)
+                L_k = np.array(L_k[:n_eig], dtype="float")
+                Y_k = np.array(Y_k[:, :n_eig], dtype="float")
+            else:  # Pre-allocate space to receive
+                L_k = np.zeros(n_eig, dtype="float")
+                Y_k = np.zeros((dim_S, n_eig), dtype="float")
+            # Broadcast smallest n_eig estimates to all ranks to compute residuals
+            self.comm.Bcast([L_k, MPI.DOUBLE], root=0)
```

Comment:
please device a global variable call MPI_MASTER_RANK to avoid using magic number `root=0`